cmds/ls/fileinfo_freebsd.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2017 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"os/user"
+	"regexp"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	humanize "github.com/dustin/go-humanize"
+)
+
+// From Linux header: /include/uapi/linux/kdev_t.h
+const (
+	minorBits = 8
+	minorMask = (1 << minorBits) - 1
+)
+
+// Matches characters which would interfere with ls's formatting.
+var unprintableRe = regexp.MustCompile("[[:cntrl:]\n]")
+
+// Since `os.FileInfo` is an interface, it is difficult to tweak some of its
+// internal values. For example, replacing the starting directory with a dot.
+// `extractImportantParts` populates our own struct which we can modify at will
+// before printing.
+type fileInfo struct {
+	name         string
+	mode         os.FileMode
+	major, minor uint64
+	uid, gid     uint32
+	size         int64
+	modTime      time.Time
+	symlink      string
+}
+
+func extractImportantParts(n string, fi os.FileInfo) fileInfo {
+	var link string
+
+	s := fi.Sys().(*syscall.Stat_t)
+	if fi.Mode()&os.ModeType == os.ModeSymlink {
+		if l, err := os.Readlink(n); err != nil {
+			link = err.Error()
+		} else {
+			link = l
+		}
+	}
+
+	return fileInfo{
+		name:    fi.Name(),
+		mode:    fi.Mode(),
+		major:   uint64(s.Rdev >> minorBits),
+		minor:   uint64(s.Rdev & minorMask),
+		uid:     s.Uid,
+		gid:     s.Gid,
+		size:    fi.Size(),
+		modTime: fi.ModTime(),
+		symlink: link,
+	}
+}
+
+// Without this cache, `ls -l` is orders of magnitude slower.
+var (
```

Comment:
if this is common to many unix-like systems maybe we need fileinfo_unix.go so we're not copy pasting code everywhere.

Reply:
Will try to abstract this out.... Though I don't understand why this doesn't just use standard FileInfo.

Reply:
Done.